pkg/controller/pipelinerun/controller_test.go

```diff
@@ -0,0 +1,175 @@
+	logging.InitializeLogger(false)
+	client := fakekubeclientset.NewSimpleClientset()
+	pipelineClient := clientset.New(&fakerestclientset.RESTClient{})
+
+	prs := []*v1alpha1.PipelineRun{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pipeline-run-success",
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.PipelineRunSpec{
+			PipelineRef: v1alpha1.PipelineRef{
+				Name:       "test-pipeline",
+				APIVersion: "a1",
+			},
+		}},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-pipeline-run-fail",
+				Namespace: "foo",
+			},
+			Spec: v1alpha1.PipelineRunSpec{
+				PipelineRef: v1alpha1.PipelineRef{
+					Name:       "test-pipeline-does-not-exist",
+					APIVersion: "a1",
+				},
+			}},
+	}
+
+	ps := []*v1alpha1.Pipeline{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pipeline",
+			Namespace: "foo",
+		}},
+	}
+
+	tcs := []struct {
+		name           string
+		pipelineRun    string
+		runtimeSuccess bool
+		expectedRecord string
+	}{
+		{"success", "foo/test-pipeline-run-success", true, "Normal Synced Found pipeline test-pipeline"},
+		{"fail", "test-pipeline-run-fail", false, ""},
+	}
+	ch := make(chan string, 10)
+	defer close(ch)
+	setUpMocks(prs, ps, ch)
+
+	c := NewController(client, pipelineClient, logging.GetContextLogger(t.Name()).Logger).(*Controller)
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			old := runtime.ErrorHandlers
+			var actualErr error
+			defer func() { runtime.ErrorHandlers = old }()
+			runtime.ErrorHandlers = []func(error){
+				func(err error) {
+					actualErr = err
+				},
+			}
+			c.workqueue.Add(tc.pipelineRun)
+			if !c.processNextWorkItem() {
+				t.Errorf("expected to process work item successfuly")
+			}
+			if tc.runtimeSuccess != (actualErr == nil) {
+				t.Errorf("expected to see error %t. Got error %v", tc.runtimeSuccess, actualErr)
+			}
+			if actualErr == nil {
+				// Check message is recorded.
+				msg := <-ch
+				if msg != tc.expectedRecord {
+					t.Errorf("expected to see %s event got %s", tc.expectedRecord, msg)
+				}
+			}
+		})
+	}
+	c.workqueue.ShutDown()
+}
+
+// sets up all the mocks for testing
+func setUpMocks(prs []*v1alpha1.PipelineRun, ps []*v1alpha1.Pipeline, ch chan string) {
+	runsListerFunc = func(i pi.PipelineRunInformer) listers.PipelineRunLister {
+		return &mockPipelineRunsLister{runs: prs}
+	}
+	pipelinesListerFunc = func(i pi.PipelineInformer) listers.PipelineLister {
+		return &mockPipelinesLister{p: ps}
+	}
+	getRecordFunc = func(k kubernetes.Interface, l *zap.SugaredLogger) record.EventRecorder {
+		return &record.FakeRecorder{
+			Events: ch,
+		}
+	}
+}
+
+type fakeStruct struct {
+}
+
+type mockPipelineRunsLister struct {
+	runs []*v1alpha1.PipelineRun
+}
+
+func (m *mockPipelineRunsLister) List(selector labels.Selector) (ret []*v1alpha1.PipelineRun, err error) {
+	return m.runs, nil
+}
+
+func (m *mockPipelineRunsLister) PipelineRuns(namespace string) listers.PipelineRunNamespaceLister {
+	return m
+}
```

Comment:
id almost rather see these all inline in `setupMocks` (if we stick to that style), it's kind of confusing to see them all just hanging here. 

either that or docstrings for each explaining what they are